### ISSUE TITLE
 App crashes on clicking on replying thread after deleting the workspace

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#19360]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to staging dot on chrome and login with User 1
2. Create a new workspace and change the name of the workspace to identify easily
3. Invite User 2
4. Go to announce of the workspace and send any message
5. Next login with User 2 on another web chrome and go to same announce
6. Click on reply on thread and send any message
7. Now come back to User 1 Account , go to #announce room and now delete the workspace
8. After deleting the workspace, click on the reply option that User 2 replied earlier
9. Result : See that the app crashed

## Expected Result:
App should not crash on clicking on replying thread after deleting the workspace
## Actual Result:
App crashes on clicking on replying thread after deleting the workspace

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.16-6
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/6d0682d2-4cfc-4fb0-b92a-da745cb5367f

https://github.com/Expensify/App/assets/43996225/ad3ebf62-dc41-4ed5-9e6d-756b4c573de8

**Expensify/Expensify Issue URL:**
**Issue reported by:** @avi-shek-jha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1684385061451199

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by dylanexpensify** [comment]
Reviewing now!

**Comment by DanutGavrus** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
App crashes on clicking on replying thread after deleting the workspace.

### What is the root cause of that problem?
After deleting the workspace, the corresponding `policy` in the `Onyx` collection inside `ReportUtils.js` gets set to __null__. So, ```const policy = lodashGet(allPolicies, `${ONYXKEYS.COLLECTION.POLICY}${report.policyID}`, {})``` sets policy to __null__ instead of `{}` which is the default value specified as a parameter, and then ```return policy.role === CONST.POLICY.ROLE.ADMIN``` crashes the app.  

Reloading the page after deleting the workspace and before pressing on the `Reply` button does not crash the app because `allPolicies` will re-initialize without the __null__ one.

### What changes do you think we should make in order to solve the problem?
I would suggest updating this:
```
ReportUtils.js
82    let allPolicies;
83    Onyx.connect({
84        key: ONYXKEYS.COLLECTION.POLICY,
85        waitForCollectionCallback: true,
86        callback: (val) => (allPolicies = val),
});
```
to this:
```
ReportUtils.js
82    let allPolicies = {};
83    Onyx.connect({
84        key: ONYXKEYS.COLLECTION.POLICY,
85        callback: (val, key) => {
86            if (!val || !key) return;
87    
88            allPolicies[key] = val;
89        },
90    });
```
No other changes are required because now: ```const policy = lodashGet(allPolicies, `${ONYXKEYS.COLLECTION.POLICY}${report.policyID}`, {})``` will set policy to `{}` and the app won't crash anymore.  

This would be better because `allPolicies` is used in many places inside `ReportUtils.js` and leaving that policy set to __null__ inside `allPolicies` may cause other similar bugs too.

**Comment by dylanexpensify** [comment]
[I'm heading OOO tomorrow](https://expensify.slack.com/archives/C01SKUP7QR0/p1684776773511559) for about a week so adding another BZ member while I'm away to help keep the train moving! 🚂

**Comment by tienifr** [comment]
I think we should not remove `waitForCollectionCallback: true` because we use it whenever we want to get the collections

**Comment by DanutGavrus** [comment]
Without removing it, we may just __filter__ out the __null__ policies before calling `allPolicies = val`

**Comment by Christinadobrzyn** [comment]
Not overdue, still working on a solution

**Comment by DanutGavrus** [comment]
@aimane-chnaif @roryabraham with that proposal the __null__ `policy` still remains in `allPolicies`. Any other `lodashGet allPolicies` will search in a list with a __null__ policy in it, possibly causing other crashes. With that fix we're just _avoiding a crash in this specific line of code_ instead of cleaning `allPolicies`.

Is it the desired behavior here?

**Comment by Christinadobrzyn** [comment]
@roryabraham what do you think of @aimane-chnaif's suggestion? https://github.com/Expensify/App/issues/19360#issuecomment-1570491123

**Comment by roryabraham** [comment]
I approve of @tienifr's proposal to fix this specific issue, but I think long-term it is pretty weird that instead of actually deleting stuff from Onyx, we just set it to null. It has caused a _lot_ of problems over the years imo. But that's a larger change that requires broader discussion than this issue warrants.

**Comment by Christinadobrzyn** [comment]
Hired in upwork - https://www.upwork.com/ab/applicants/1660550422417498112/job-details

- @aimane-chnaif = C+
- @tienifr = contributor
- @avi-shek-jha = reporter

**Comment by avi-shek-jha** [comment]
@Christinadobrzyn Submitted the work. Thanks

**Comment by Christinadobrzyn** [comment]
Moved to weekly while we review the PR

**Comment by cead22** [comment]
From @avi-shek-jha [here](https://github.com/Expensify/App/issues/20122#issuecomment-1579863117)

>Seems like the app crash issue has been solved. But still there's a second error that has arisen. When removed the user from the workspace, the chat page still remains and a removed User B is able to reply through the chat composer. The chat page should have been disappeared in a normal scenario like how announce channel gets disappeared if a user has been removed. I'm not sure but may be a regression from the 19360 issue.

![Screenshot from 2023-06-07 10-08-11](https://github.com/Expensify/App/assets/123831727/369112e4-2f8a-4d6a-ad6d-eee2c6a46658)


@avi-shek-jha can you confirm if that was supposed to be fixed along with this issue?

**Comment by avi-shek-jha** [comment]
Hi @cead22. I don't have much idea on this. But, if you see in both the issue #20122 and #19360, the main bug was the app crashing issue that needed to be solved. However, what I noticed is when the #19360 was solved, the app crashing issue on #20122 was also solved. But a new issue still remains on #20122 that still needs to be fixed. Not sure, whether it should have been solved here or not.

 @roryabraham @Christinadobrzyn Can you please look into this once. thanks.

**Comment by Christinadobrzyn** [comment]
Today is the day for payment! 

- No regressions so paying this out. 
- Looks like this job is not eligible for a speed bonus (PR created May 31st & merged June 5th)
- Paid out based on this comment - https://github.com/Expensify/App/issues/19360#issuecomment-1577534678

I think we can skip a regression test for this? @aimane-chnaif

**Comment by dylanexpensify** [comment]
@Christinadobrzyn did you issue payment? If not I can tackle!

**Comment by Christinadobrzyn** [comment]
Sorry not paid yet, paying this now.

**Comment by Christinadobrzyn** [comment]
Okay I can't add a bonus to the original job but I created this one for the bonus to @aimane-chnaif and @tienifr 

Upwork job internal: https://www.upwork.com/ab/applicants/1669434505911701505/job-details
Upwork job external: https://www.upwork.com/jobs/~012cd838303967343c

Can you please apply to this so I can pay out the bonus?

**Comment by blockchain1112** [comment]
Hello @Christinadobrzyn

**Comment by blockchain1112** [comment]
Hello @Christinadobrzyn

**Comment by kooya3** [comment]
**After deleting the workspace, the corresponding policy in the Onyx collection inside ReportUtils.js gets set to null. This causes the crash when trying to access the null policy.
my proposal is to update ReportUtils.js to handle the null policy case and prevent the app from crashing.**

**Comment by kooya3** [comment]
Contributor details
Your Expensify account email: elyees.dev@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01f35492ac1d67b51a

**Comment by kooya3** [comment]
# Fixing Form Submission Bug: Proposal for Correcting Data Processing Issue

## Overview
The form submission bug causes incorrect data processing on the server side, leading to inaccurate and incomplete data storage within the web application. This impacts the application's functionality and user experience.

## Objective
The objective of this proposal is to address the form submission bug and ensure that the submitted data is processed accurately and completely.

## Proposed Solution

### Approach
The proposed solution involves analyzing and resolving the root cause of the bug in the form submission process. This will require a thorough examination of the code responsible for handling form submissions and making necessary modifications.

### Technical Implementation
- Analyze the server-side code responsible for processing form submissions: Review the relevant code files and functions/methods related to form submission handling.
- Identify the specific function or method where the bug occurs: Identify the section of code that is responsible for the incorrect data processing.
- Debug the code to understand the reason behind the incorrect data processing: Step through the code and use debugging tools to identify the underlying cause of the bug.
- Implement the necessary fixes, ensuring that the submitted data is validated, processed correctly, and stored accurately in the application's database: Make the required code modifications to address the bug, ensuring proper validation and accurate data processing.

### Testing Plan
- Develop a suite of unit tests to verify the correct functioning of the form submission process: Create tests that cover different scenarios, including valid and invalid data submissions, to ensure the bug fix is effective.
- Execute integration tests to ensure the bug fix does not introduce any regressions or conflicts with other application components: Verify that the fixed form submission process works seamlessly with other parts of the application.

### Potential Risks and Mitigation
- Note the possibility of introducing new bugs or unintended consequences while modifying the code: Acknowledge the potential risks associated with code changes and emphasize the need for thorough testing and code reviews to mitigate them.

### Timeline and Deliverables
- Task 1: Code analysis and identification of the bug (2 days).
- Task 2: Debugging and implementing the fix (3 days).
- Task 3: Writing unit tests and conducting integration testing (2 days).
- Task 4: Bug fixes and final code review (1 day).

### Resource Requirements
- Access to the codebase and development environment: Request necessary permissions and credentials to access and modify the codebase.
- Collaboration with the backend development team, if applicable: Coordinate with the backend team if the bug involves server-side code changes that require their expertise.

### Communication and Collaboration
- Weekly progress updates via email or project management tools: Commit to regular communication to keep the client and stakeholders informed of progress and any challenges faced.
- Open communication channels for addressing any questions or concerns: Provide contact information and establish a communication protocol for effective collaboration.

## Next Steps
Seek feedback on the proposal, gain approval to proceed, and initiate the bug fixing process: Share the proposal with the client or relevant stakeholders, gather their input, and obtain approval before starting the bug fixing process.

**Comment by Christinadobrzyn** [comment]
Hi @blockchain1112 and @kooya3 - this job is complete but if you'd like to work on another we have listed open jobs here (look for the ones with Help Wanted & External label) - https://github.com/Expensify/App/issues?q=is%3Aissue+is%3Aopen+help+wanted+label%3A%22Help+Wanted%22+label%3AExternal

**Comment by blockchain1112** [comment]
Okay Good So what should I do?
